File: testing/robertwb-python/strassen.py
from random import xorshift128plus as rand
import copy


def generateMatrix(rows, columns, mask):
    matrix = []
    for i in range(rows):
        row = []
        for j in range(columns):
            row.append(rand() & mask)
        matrix.append(row)
    return matrix


def printMatrix(matrix):
    for row in matrix:
        for val in row:
            print('{:10}'.format(val), end="")
        print()


def matrixMultiplication(A, B):
    # A = nxm, B = mxp, C = nxp
    n = len(A)
    m = len(A[0])
    p = len(B[0])
    C = [[0]*p for i in range(n)]
    for i in range(n):
        for j in range(p):
            theSum = 0
            for k in range(m):
                theSum += A[i][k] * B[k][j]
            C[i][j] = theSum
    return C


def addMatrix(A, B):
    C = copy.deepcopy(A)
    for i in range(len(C)):
        for j in range(len(C[i])):
            C[i][j] += B[i][j]
    return C


def subtractMatrix(A, B):
    C = copy.deepcopy(A)
    for i in range(len(C)):
        for j in range(len(C[i])):
            C[i][j] -= B[i][j]
    return C


def strassen(A, B):
    if len(A) <= 128:
        return matrixMultiplication(A, B)
    mid = len(A)//2
    A11 = [A[i][:mid] for i in range(0, mid)]
    A12 = [A[i][mid:] for i in range(0, mid)]
    A21 = [A[i][:mid] for i in range(mid, len(A))]
    A22 = [A[i][mid:] for i in range(mid, len(A))]
    # print()
    # printMatrix(A11)
    # print()
    # printMatrix(A12)
    # print()
    # printMatrix(A21)
    # print()
    # printMatrix(A22)
    B11 = [B[i][:mid] for i in range(0, mid)]
    B12 = [B[i][mid:] for i in range(0, mid)]
    B21 = [B[i][:mid] for i in range(mid, len(B))]
    B22 = [B[i][mid:] for i in range(mid, len(B))]
    S1 = addMatrix(A21, A22)
    S2 = subtractMatrix(S1, A11)
    S3 = subtractMatrix(A11, A21)
    T1 = subtractMatrix(B12, B11)
    T2 = subtractMatrix(B22, T1)
    T3 = subtractMatrix(B22, B12)
    S4 = subtractMatrix(A12, S2)
    T4 = subtractMatrix(T2, B21)
    P1 = strassen(A11, B11)
    P2 = strassen(A12, B21)
    P3 = strassen(S4, B22)
    P4 = strassen(A22, T4)
    P5 = strassen(S1, T1)
    P6 = strassen(S2, T2)
    P7 = strassen(S3, T3)
    U1 = addMatrix(P1, P2)
    U2 = addMatrix(P1, P6)
    U3 = addMatrix(U2, P7)
    U4 = addMatrix(U2, P5)
    U5 = addMatrix(U4, P3)
    U6 = subtractMatrix(U3, P4)
    U7 = addMatrix(U3, P5)

    # printMatrix(U1)
    # print()
    # printMatrix(U6)
    # print()
    # printMatrix(U5)
    # print()
    # printMatrix(U7)

    U1.extend(U6)
    U5.extend(U7)

    # print()
    # printMatrix(U1)
    # print()
    # printMatrix(U5)
    # print()
    for i in range(len(U1)):
        U1[i].extend(U5[i])
    # printMatrix(U1)
    return U1


def main():
    print(rand())
    m1 = generateMatrix(512, 512, 10)
    m2 = generateMatrix(512, 512, 10)
    # print()
    # printMatrix(m1)
    # print()
    # printMatrix(m2)
    res = matrixMultiplication(m1, m2)
    printMatrix(res)
    # C = strassen(m1, m2)
    # printMatrix(C)

if __name__ == '__main__':
    main()

File: testing/robertwb-c/strassen.c
#include "strassen.h"
void strassen(uint64_t **A, uint64_t **B, uint64_t **C, size_t size) {
  if (size < 128) {
    matrixMultiplicationFixed(A, B, C);
    return;
  }
  size_t mid = size/2;
  uint64_t **A11 = getArray(mid);
  uint64_t **A12 = getArray(mid);
  uint64_t **A21 = getArray(mid);
  uint64_t **A22 = getArray(mid);
  uint64_t **B11 = getArray(mid);
  uint64_t **B12 = getArray(mid);
  uint64_t **B21 = getArray(mid);
  uint64_t **B22 = getArray(mid);

  for (size_t i = 0; i < mid; i++) {
    for (size_t j = 0; j < mid; j++) {
      A11[i][j] = A[i][j];
      A12[i][j] = A[i][j+mid];
      A21[i][j] = A[i+mid][j];
      A22[i][j] = A[i+mid][j+mid];
      B11[i][j] = B[i][j];
      B12[i][j] = B[i][j+mid];
      B21[i][j] = B[i+mid][j];
      B22[i][j] = B[i+mid][j+mid];
    }
  }

  uint64_t **S1 = getArray(mid);
  uint64_t **S2 = getArray(mid);

  addNew(A21, A22, S1, mid);
  subNew(S1, A11, S2, mid);

  subRight(A11, A21, mid);
  uint64_t **S3 = A21;

  uint64_t **T1 = getArray(mid);
  uint64_t **T2 = getArray(mid);
  subNew(B12, B11, T1, mid);
  subNew(B22, T1, T2, mid);

  subRight(B22, B12, mid);
  uint64_t **T3 = B12;

  uint64_t **S4 = getArray(mid);
  uint64_t **T4 = getArray(mid);
  subNew(A12, S2, S4, mid);
  subNew(T2, B21, T4, mid);

  uint64_t **P1 = getArray(mid);
  strassen(A11, B11, P1, mid);
  free(A11[0]);
  free(A11);
  strassen(A12, B21, B11, mid);
  uint64_t **P2 = B11;
  strassen(S4, B22, B21, mid);
  uint64_t **P3 = B21;
  strassen(A22, T4, A12, mid);
  free(A22[0]);
  free(A22);
  uint64_t **P4 = A12;
  strassen(S1, T1, S4, mid);
  free(T1[0]);
  free(T1);
  free(S1[0]);
  free(S1);
  uint64_t **P5 = S4;
  strassen(S2, T2, B22, mid);
  free(S2[0]);
  free(S2);
  free(T2[0]);
  free(T2);
  uint64_t **P6 = B22;
  strassen(S3, T3, T4, mid);
  uint64_t **P7 = T4;
  free(S3[0]);
  free(S3);
  free(T3[0]);
  free(T3);

  //A22, T1, S1, S2, T2, S3, T3,
  //P6, P3
  addLeft(P2, P1, mid);
  uint64_t **U1 = P2;
  addLeft(P1, P6, mid);
  free(P6[0]);
  free(P6);
  uint64_t **U2 = P1;
  addLeft(P7, U2, mid);
  uint64_t **U3 = P7;
  addLeft(U2, P5, mid);
  uint64_t **U4 = U2;
  addLeft(U4, P3, mid);
  free(P3[0]);
  free(P3);
  uint64_t **U5 = U4;
  subRight(U3, P4, mid);
  uint64_t **U6 = P4;
  addLeft(U3, P5, mid);
  uint64_t **U7 = U3;
  free(P5[0]);
  free(P5);

  for (size_t i = 0; i < mid; i++) {
    for (size_t j = 0; j < mid; j++) {
      C[i][j] = U1[i][j];
      C[i][j+mid] = U5[i][j];
      C[i+mid][j] = U6[i][j];
      C[i+mid][j+mid] = U7[i][j];
    }
  }
  free(U1[0]);
  free(U1);
  free(U5[0]);
  free(U5);
  free(U6[0]);
  free(U6);
  free(U7[0]);
  free(U7);

  return;
}
